src/Microsoft.VisualStudio.Threading.Tests/AsyncReaderWriterLockTests.cs

```diff
@@ -1813,6 +1813,49 @@ public void WriteLockAsyncSynchronousReleaseAllowsOtherWriters()
             Assert.True(firstLockTask.Wait(TestTimeout)); // rethrow any exceptions
         }
 
+        [StaFact]
+        public void WriteLockDisposingShouldNotBlockByOtherWriters()
+        {
+            var firstLockToRelease = new AsyncManualResetEvent();
+            var firstLockAccquired = new AsyncManualResetEvent();
+            var firstLockToDispose = new AsyncManualResetEvent();
+            var firstLockTask = Task.Run(async delegate
+            {
+                using (var firstLock = await this.asyncLock.WriteLockAsync())
+                {
+                    firstLockAccquired.Set();
+                    await firstLockToRelease.WaitAsync();
+                    await firstLock.ReleaseAsync();
+                    await firstLockToDispose.WaitAsync();
+                }
+            });
+
+            var secondLockReleased = new AsyncManualResetEvent();
+
+            var secondLockTask = Task.Run(async delegate
+            {
+                await firstLockAccquired.WaitAsync();
+                var awaiter = this.asyncLock.WriteLockAsync().GetAwaiter();
+                Assert.False(awaiter.IsCompleted);
+                awaiter.OnCompleted(() =>
+                {
+                    firstLockToDispose.Set();
+
+                    // We must block the thread synchorizely, so it won't release the NonConcurrentSynchronizationContext.
+                    firstLockTask.Wait();
+
+                    awaiter.GetResult().Dispose();
```

Comment:
This test would more accurately reflect real code if the very first thing the OnCompleted delegate did was call `awaiter.GetResult()`. You could store it in a local variable and dispose it at this point. But calling it *after* synchronously blocking would never happen in a natural `await` situation so it makes the test appear too artificial, as if there isn't a real problem that it's testing for.

Reply:
fixed.